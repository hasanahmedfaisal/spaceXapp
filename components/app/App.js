/* global fetch */
import React from 'react'
import isEmpty from 'lodash/isEmpty'
import Card from '../card/Card'
import CardFilter from '../cardFilter/CardFilter'
import getRequestDetails from '../../helpers/RequestHelper'
import { parse } from '../../helpers/ResultsParser'

class App extends React.Component {
  constructor (props) {
    super(props)
    this.state = {
      filters: props.filters,
      data: props.data,
      isLoading: false,
      isError: false
    }
    this.fetchData = this.fetchData.bind(this)
  }

  componentDidMount () {
    if (!isEmpty(this.props.filters)) {
      this.fetchData(this.props.filters)
    }
  }

  async onFilterChange (selectedFilters) {
    await this.fetchData(selectedFilters)
  }

  async fetchData (selectedFilters) {
    try {
      this.setState({
        isLoading: true
      })
      const res = await fetch(getRequestDetails(selectedFilters).endpoint)
      const result = await res.json()
      const data = parse(result)
      this.setState({
        data,
        filters: selectedFilters,
        isLoading: false
      })
    } catch (error) {
      this.setState({
        isError: true
      })
    }
  }

  render () {
    return (
      <div className='App'>
        <div className='title'>SpaceX Launch Programs</div>
        <div className='mainContainer'>
          <CardFilter filters={this.state.filters} handleClick={this.onFilterChange} />
          {this.state.isLoading && <div className='loader'>Loading...</div>}
          {!isEmpty(this.state.data) && !this.state.isLoading &&
            <div className='cardContainer'>
              {this.state.data.map((item, index) => (
                <Card key={index} cardProps={item} />
              ))}
            </div>}
        </div>
      </div>
    )
  }
}

export default App
