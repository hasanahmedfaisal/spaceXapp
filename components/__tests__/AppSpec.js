import App from '../app/App'
import renderer from 'react-test-renderer'

describe('App',() => {
    it('should render correctly',() => {
        const appProps = {
            data:[{
             mission_name:'some mission name',
             flight_number: 'some flight name', 
             launch_year: 'some launch year', 
             launch_success: true, 
             land_success: true, 
             mission_patch_small: 'some url', 
             mission_id: ['123']
            }],
            filters: {
                "launch_year": '2010'
            }
        }
        const tree = renderer.create(<App {...appProps}/>).toJSON()
        expect(tree).toMatchSnapshot()
    }) 
})