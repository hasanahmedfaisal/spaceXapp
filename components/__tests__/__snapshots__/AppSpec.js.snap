// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render correctly when filters are present 1`] = `
<div
  className="App"
>
  <div
    className="title"
  >
    SpaceX Launch Programs
  </div>
  <div
    className="mainContainer"
  >
    <div
      className="mainCardContainer"
    >
      <div
        className="filters"
      >
        Filters
      </div>
      <div
        className="launchYear"
      >
        Launch Year
      </div>
      <div
        className="buttonContainer"
      >
        <div
          className="buttonDiv"
        >
          <button
            className="button"
            onClick={[Function]}
          >
            2006
          </button>
        </div>
        <div
          className="buttonDiv"
        >
          <button
            className="button"
            onClick={[Function]}
          >
            2007
          </button>
        </div>
        <div
          className="buttonDiv"
        >
          <button
            className="button"
            onClick={[Function]}
          >
            2008
          </button>
        </div>
        <div
          className="buttonDiv"
        >
          <button
            className="button"
            onClick={[Function]}
          >
            2009
          </button>
        </div>
        <div
          className="buttonDiv"
        >
          <button
            className="button isSelected"
            onClick={[Function]}
          >
            2010
          </button>
        </div>
        <div
          className="buttonDiv"
        >
          <button
            className="button"
            onClick={[Function]}
          >
            2011
          </button>
        </div>
        <div
          className="buttonDiv"
        >
          <button
            className="button"
            onClick={[Function]}
          >
            2012
          </button>
        </div>
        <div
          className="buttonDiv"
        >
          <button
            className="button"
            onClick={[Function]}
          >
            2013
          </button>
        </div>
        <div
          className="buttonDiv"
        >
          <button
            className="button"
            onClick={[Function]}
          >
            2014
          </button>
        </div>
        <div
          className="buttonDiv"
        >
          <button
            className="button"
            onClick={[Function]}
          >
            2015
          </button>
        </div>
        <div
          className="buttonDiv"
        >
          <button
            className="button"
            onClick={[Function]}
          >
            2016
          </button>
        </div>
        <div
          className="buttonDiv"
        >
          <button
            className="button"
            onClick={[Function]}
          >
            2017
          </button>
        </div>
        <div
          className="buttonDiv"
        >
          <button
            className="button"
            onClick={[Function]}
          >
            2018
          </button>
        </div>
        <div
          className="buttonDiv"
        >
          <button
            className="button"
            onClick={[Function]}
          >
            2019
          </button>
        </div>
        <div
          className="buttonDiv"
        >
          <button
            className="button"
            onClick={[Function]}
          >
            2020
          </button>
        </div>
      </div>
      <div
        className="launchYear"
      >
        Successful Launch
      </div>
      <div
        className="buttonContainer"
      >
        <div
          className="buttonDiv"
        >
          <button
            className="button"
            onClick={[Function]}
          >
            True
          </button>
        </div>
        <div
          className="buttonDiv"
        >
          <button
            className="button"
            onClick={[Function]}
          >
            False
          </button>
        </div>
      </div>
      <div
        className="launchYear"
      >
        Successful Landing
      </div>
      <div
        className="buttonContainer"
      >
        <div
          className="buttonDiv"
        >
          <button
            className="button"
            onClick={[Function]}
          >
            True
          </button>
        </div>
        <div
          className="buttonDiv"
        >
          <button
            className="button"
            onClick={[Function]}
          >
            False
          </button>
        </div>
      </div>
    </div>
    <div
      className="loader"
    >
      Loading...
    </div>
  </div>
</div>
`;

exports[`App should render correctly when no filters are present 1`] = `
<div
  className="App"
>
  <div
    className="title"
  >
    SpaceX Launch Programs
  </div>
  <div
    className="mainContainer"
  >
    <div
      className="mainCardContainer"
    >
      <div
        className="filters"
      >
        Filters
      </div>
      <div
        className="launchYear"
      >
        Launch Year
      </div>
      <div
        className="buttonContainer"
      >
        <div
          className="buttonDiv"
        >
          <button
            className="button"
            onClick={[Function]}
          >
            2006
          </button>
        </div>
        <div
          className="buttonDiv"
        >
          <button
            className="button"
            onClick={[Function]}
          >
            2007
          </button>
        </div>
        <div
          className="buttonDiv"
        >
          <button
            className="button"
            onClick={[Function]}
          >
            2008
          </button>
        </div>
        <div
          className="buttonDiv"
        >
          <button
            className="button"
            onClick={[Function]}
          >
            2009
          </button>
        </div>
        <div
          className="buttonDiv"
        >
          <button
            className="button"
            onClick={[Function]}
          >
            2010
          </button>
        </div>
        <div
          className="buttonDiv"
        >
          <button
            className="button"
            onClick={[Function]}
          >
            2011
          </button>
        </div>
        <div
          className="buttonDiv"
        >
          <button
            className="button"
            onClick={[Function]}
          >
            2012
          </button>
        </div>
        <div
          className="buttonDiv"
        >
          <button
            className="button"
            onClick={[Function]}
          >
            2013
          </button>
        </div>
        <div
          className="buttonDiv"
        >
          <button
            className="button"
            onClick={[Function]}
          >
            2014
          </button>
        </div>
        <div
          className="buttonDiv"
        >
          <button
            className="button"
            onClick={[Function]}
          >
            2015
          </button>
        </div>
        <div
          className="buttonDiv"
        >
          <button
            className="button"
            onClick={[Function]}
          >
            2016
          </button>
        </div>
        <div
          className="buttonDiv"
        >
          <button
            className="button"
            onClick={[Function]}
          >
            2017
          </button>
        </div>
        <div
          className="buttonDiv"
        >
          <button
            className="button"
            onClick={[Function]}
          >
            2018
          </button>
        </div>
        <div
          className="buttonDiv"
        >
          <button
            className="button"
            onClick={[Function]}
          >
            2019
          </button>
        </div>
        <div
          className="buttonDiv"
        >
          <button
            className="button"
            onClick={[Function]}
          >
            2020
          </button>
        </div>
      </div>
      <div
        className="launchYear"
      >
        Successful Launch
      </div>
      <div
        className="buttonContainer"
      >
        <div
          className="buttonDiv"
        >
          <button
            className="button"
            onClick={[Function]}
          >
            True
          </button>
        </div>
        <div
          className="buttonDiv"
        >
          <button
            className="button"
            onClick={[Function]}
          >
            False
          </button>
        </div>
      </div>
      <div
        className="launchYear"
      >
        Successful Landing
      </div>
      <div
        className="buttonContainer"
      >
        <div
          className="buttonDiv"
        >
          <button
            className="button"
            onClick={[Function]}
          >
            True
          </button>
        </div>
        <div
          className="buttonDiv"
        >
          <button
            className="button"
            onClick={[Function]}
          >
            False
          </button>
        </div>
      </div>
    </div>
    <div
      className="cardContainer"
    >
      <div
        className="mainCard"
      >
        <img
          alt="Image Unavailable"
          src="some url"
        />
        <div
          className="text missionName"
        >
          some mission name
           #
          some flight name
        </div>
        <div
          className="text missionId"
        >
          Mission Ids:
          <div
            className="missionIds"
          >
            123
          </div>
        </div>
        <div
          className="text launchYr"
        >
          Launch Year: 
          some launch year
        </div>
        <div
          className="text launchStatus"
        >
          Successful Launch: 
          True
        </div>
        <div
          className="text landingStatus"
        >
          Successful Landing: 
          True
        </div>
      </div>
    </div>
  </div>
</div>
`;
